### DISCUSSION TITLE
Berry interpreter library

### DISCUSSION THREAD

**Discussion by Sebight** [berry-lang/berry#425] [Q&A] [answered]
Hello, I'm quite confused about the actual Berry interpreter code and whether it can be used in an existing ESP32 firmware as a library to execute additional Berry code.

Could someone please point me somewhere? Thanks a lot.

**Comment by diazvictor** [comment]
Tasmota supports ESP32

https://tasmota.github.io/docs/ESP32/#berry-scripting

**Reply by Sebight** [reply]
Yes I know, but I am asking (sorry, was not probably clear in my original post) whether I can use it outside of Tasmota in my own ESP32 firmware.

In the meantime, I found some additional sources, such as this one: https://github.com/HoGC/esp32_berry. However, upon code compilation, I get a lot of errors in the Berry codebase (be_solidifylib.c, ...) failing because of **-Werror=char-subscripts** and other similar errors.

Could I find some sort of documentation on implementing the Berry interpreter into my own codebase? Thank you!

**Reply by s-hadinger** [reply]
It's hard to get a very general documentation for this, but it's not too hard. What you need is here:

https://berry.readthedocs.io/en/latest/source/en/Chapter-9.html

and here:

https://berry.readthedocs.io/en/latest/source/en/Appendix-C.html

**Reply by Sebight** [reply]
Thanks for the resources. I checked some of these earlier and I think the biggest problem for me was the lack of documentation in terms of the **coc** command and all of that.

In other words, I'm confident I would find a way around the C++ code, but the problem is the part before that, as in getting the library to load all stuff correctly and compile.

**Reply by s-hadinger** [reply]
Ah ok. I rewrote the `coc` in python to make it easier to integrate in a CI. Don't judge the code quality, it's a direct port from C++, so not very pythonic.

You can trigger the `coc` part with `make prebuild` and then do a normal C/C++ compilation

**Reply by Sebight** [reply]
Thanks a lot. I will look into it and see if I have any success.

**Reply by Sebight** [reply]
So running the `make prebuild` will create the `generate` folder. That's one of the folders I want aside from the main berry source code. 

![image](https://github.com/berry-lang/berry/assets/59232422/5fe0b7d2-42ff-4e75-a1fb-8227416a9795)

Is the `port` folder autogenerated, or do I have to create it manually?

Thank you!

**Reply by s-hadinger** [reply]
What's in `port` needs to be written or changed manually

**Reply by Sebight** [reply]
Hello, thanks. I tried it out, and the same problems are still persisting.

Let me break it down:

## Project structure
`berry/` is the cloned repository.
![image](https://github.com/berry-lang/berry/assets/59232422/c7f8a913-93ad-4c11-8c8e-02b895e7b951)

Everything is referred to in this CMake:
```
file(GLOB berry_srcs
          "port/*.c"
          "berry/src/*.c")

idf_component_register(SRCS ${berry_srcs}
                    INCLUDE_DIRS "port" "generate" "berry/src")
```

The good news is that everything compiles! Or at least all the code until it errors.

When I compile it, I start getting compile errors from Berry's internal scripts.

![image](https://github.com/berry-lang/berry/assets/59232422/ccc0c3df-0cca-4f15-9dbe-18855c541d0d)

I could probably tell the compiler not to treat these as errors. However, I don't think that's a proper solution.

Any idea what might be up? Thanks!

**Reply by Sebight** [reply, marked as answer]
Solved this by adding this to my buildflags:
`build_flags = -Wno-pointer-sign -Wno-char-subscripts`
